ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/tracing/vm/VmTraceGenerator.java

```diff
@@ -0,0 +1,284 @@
+      }
+    }
+  }
+
+  private void generateTracingStorage(final VmOperationExecutionReport report) {
+    // set storage if updated
+    updatedStorage(lastCapturedStorage, currentTraceFrame.getStorage())
+        .map(
+            storageEntry ->
+                new Store(
+                    storageEntry.key.toShortHexString(), storageEntry.value.toShortHexString()))
+        .ifPresent(report::setStore);
+  }
+
+  /**
+   * Set current trace from parents queue and retrieve next frame.
+   *
+   * @param frame the trace frame.
+   */
+  private void initStep(final TraceFrame frame) {
+    this.currentTraceFrame = frame;
+    currentTrace = parentTraces.getLast();
+    // set smart contract code
+    currentTrace.setCode(
+        currentTraceFrame.getMaybeCode().orElse(new Code()).getBytes().getHexString());
+    final int nextFrameIndex = currentIndex + 1;
+    // retrieve next frame if not last
+    maybeNextFrame =
+        transactionTrace.getTraceFrames().size() > nextFrameIndex
+            ? Optional.of(transactionTrace.getTraceFrames().get(nextFrameIndex))
+            : Optional.empty();
+  }
+
+  /**
+   * Find updated storage from 2 storage captures.
+   *
+   * @param firstCapture The first storage capture.
+   * @param secondCapture The second storage capture.
+   * @return an {@link Optional} wrapping the diff.
+   */
+  private Optional<StorageEntry> updatedStorage(
+      final Optional<Map<UInt256, UInt256>> firstCapture,
+      final Optional<Map<UInt256, UInt256>> secondCapture) {
+    final Map<UInt256, UInt256> first = firstCapture.orElse(new HashMap<>());
+    final Map<UInt256, UInt256> second = secondCapture.orElse(new HashMap<>());
+    final MapDifference<UInt256, UInt256> diff = Maps.difference(first, second);
+    final Map<UInt256, MapDifference.ValueDifference<UInt256>> entriesDiffering =
+        diff.entriesDiffering();
+    if (entriesDiffering.size() > 0) {
+      final UInt256 firstDiffKey = entriesDiffering.keySet().iterator().next();
+      final MapDifference.ValueDifference<UInt256> firstDiff = entriesDiffering.get(firstDiffKey);
+      return Optional.of(new StorageEntry(firstDiffKey, firstDiff.rightValue()));
+    }
+    final Map<UInt256, UInt256> onlyOnRight = diff.entriesOnlyOnRight();
+    if (onlyOnRight.size() > 0) {
+      final UInt256 firstOnlyOnRightKey = onlyOnRight.keySet().iterator().next();
+      return Optional.of(
+          new StorageEntry(firstOnlyOnRightKey, onlyOnRight.get(firstOnlyOnRightKey)));
+    }
+    return Optional.empty();
+  }
+
+  private Optional<Bytes32[]> updatedMemory(
+      final Optional<Bytes32[]> firstCapture, final Optional<Bytes32[]> secondCapture) {
+    final Bytes32[] first = firstCapture.orElse(new Bytes32[0]);
+    final Bytes32[] second = secondCapture.orElse(new Bytes32[0]);
+    final boolean deepEquals = Arrays.deepEquals(first, second);
+    if (deepEquals) {
+      return Optional.empty();
+    }
+    return Optional.of(second);
```

Comment:
Can we add some tests that do sequential memory modification to see if this works in those cases?  Just want to make sure that we shouldn't be generating a memory diff here.